### PR TITLE
fix @Fluent when properties are all required and in ctor args

### PR DESCRIPTION
**Pull request by XiaofeiCao**
fix https://github.com/Azure/autorest.java/issues/1654

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
you can run the action https://github.com/Azure/autorest.java/actions/workflows/generate-test-code.yml to update autorest test code

but make sure your coding is correct.

File: javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java [hunk 1]
```diff
@@ -411,6 +405,31 @@ protected void addClassLevelAnnotations(ClientModel model, JavaFile javaFile, Ja
...
+    private void addFluentOrImmutableAnnotation(ClientModel model, JavaFile javaFile,
+                                                List<ClientModelPropertyReference> propertyReferences, JavaSettings settings) {
+        boolean isImmutable =
+                // only contains read-only properties
+                (model.getProperties().stream().allMatch(ClientModelProperty::getIsReadOnly)
+                        && propertyReferences.stream().allMatch(ClientModelPropertyReference::getIsReadOnly))
+                // only contains required properties and "requiredFieldsAsConstructorArgs" is set to true
+                || (settings.isRequiredFieldsAsConstructorArgs()
+                        && model.getProperties().stream().allMatch(ClientModelProperty::isRequired)
+                        && propertyReferences.stream().allMatch(ClientModelPropertyReference::isRequired));
```
[the remark this answers is not in the document]
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java:423]: Got it. If the model contains the mix of the two kinds of properties, this logic breaks. I should check the property separately. 
Thanks, will update the test case.

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java [hunk 2]
```diff
@@ -410,15 +412,29 @@ public static List<ClientModelProperty> getRequiredParentProperties(ClientModel
...
         // add a setter.
-        return !property.getIsReadOnly()
-            && !(settings.isRequiredFieldsAsConstructorArgs() && property.isRequired())
-            && !property.getClientFlatten();
+        return !readOnlyOrInConstructor(property, settings)
+                && !privateAccess(property);
+    }
+
+    // A property has private access when it is to be flattened.
+    private static boolean privateAccess(ClientModelPropertyAccess property) {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java:426]: isPrivateAccess?

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java [hunk 3]
```diff
@@ -410,15 +412,29 @@ public static List<ClientModelProperty> getRequiredParentProperties(ClientModel
...
+        boolean privateAccess = false;
+        // ClientModelPropertyReference never refers to a private access property, so only check ClientModelProperty here.
+        if (property instanceof ClientModelProperty) {
+            privateAccess = ((ClientModelProperty) property).getClientFlatten();
+        }
+        return privateAccess;
+    }
+
+    private static boolean readOnlyOrInConstructor(ClientModelPropertyAccess property, JavaSettings settings) {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java:435]: isReadOnlyOrInConstructor?

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java [hunk 4]
```diff
@@ -410,15 +412,29 @@ public static List<ClientModelProperty> getRequiredParentProperties(ClientModel
...
+        return !readOnlyOrInConstructor(property, settings)
+                && !privateAccess(property);
+    }
+
+    // A property has private access when it is to be flattened.
+    private static boolean privateAccess(ClientModelPropertyAccess property) {
+        boolean privateAccess = false;
+        // ClientModelPropertyReference never refers to a private access property, so only check ClientModelProperty here.
+        if (property instanceof ClientModelProperty) {
+            privateAccess = ((ClientModelProperty) property).getClientFlatten();
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java:430]: Note:
This `clientFlatten` only happen in mgmt, hence no effect on vanilla or DPG.

**Review by weidongxu-microsoft — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java (hunk 2 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java]: Changed.

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java (hunk 3 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java]: Changed.

File: javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java (hunk 4 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/util/ClientModelUtil.java:431]: Thanks, added comment on this.

**Review by haolingdong-msft — APPROVED** [review]

**Pull request merged**